common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java

```diff
@@ -91,4 +91,64 @@ public static boolean arrayEquals(
     }
     return true;
   }
+
+  public static boolean contains(byte[] array, byte[] target) {
+    if (target.length == 0) {
+      return true;
+    } else if (array.length == 0) {
+      return false;
+    } else if (target.length > array.length) {
+      return false;
+    } else {
+      return indexOf(array, target) >= 0;
+    }
+  }
+
+  public static boolean startsWith(byte[] array, byte[] target) {
+    if (target.length == 0) {
+      return true;
+    } else if (array.length == 0) {
+      return false;
+    } else if (target.length > array.length) {
+      return false;
+    } else {
+      return indexOf(array, target) == 0;
+    }
+  }
+
+  public static boolean endsWith(byte[] array, byte[] target) {
+    if (target.length == 0) {
+      return true;
+    } else if (array.length == 0) {
+      return false;
+    } else if (target.length > array.length) {
+      return false;
+    } else {
+      int endPos = array.length - target.length;
+      for (int i = 0; i < target.length; ++i) {
+        if (array[endPos + i] != target[i]) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
```

Comment:
Is it possible to use optimized method `arrayEquals` in `startsWith()` and `endsWith()` as UTF8String methods do? I guess it is faster than your `indexOf()`.

Reply:
Also we need to check that there is no perf regression in UTF8String methods. Please, run string benchmarks.

Reply:
Updated with same method like `UTF8String`, so performance is not a concern?